src/reader/warming.rs

```diff
@@ -0,0 +1,341 @@
+
+impl WarmingState {
+    pub fn new(
+        num_warming_threads: usize,
+        warmers: Vec<Weak<dyn Warmer>>,
+        searcher_generation_inventory: Inventory<SearcherIndexGeneration>,
+    ) -> crate::Result<Self> {
+        Ok(Self(Arc::new(Mutex::new(WarmingStateInner {
+            num_warming_threads,
+            warmers,
+            gc_thread: None,
+            searcher_generation_ids: Default::default(),
+            searcher_generation_inventory,
+        }))))
+    }
+
+    /// Start tracking a new generation of [Searcher], and [Warmer::warm] it if there are active warmers.
+    ///
+    /// A background GC thread for [Warmer::garbage_collect] calls is uniquely created if there are active warmers.
+    pub fn warm_new_searcher_generation(&self, searcher: &Searcher) -> crate::Result<()> {
+        self.0
+            .lock()
+            .unwrap()
+            .warm_new_searcher_generation(searcher, &self.0)
+    }
+
+    #[cfg(test)]
+    fn gc_maybe(&self) -> bool {
+        self.0.lock().unwrap().gc_maybe()
+    }
+}
+
+struct WarmingStateInner {
+    num_warming_threads: usize,
+    warmers: Vec<Weak<dyn Warmer>>,
+    gc_thread: Option<JoinHandle<()>>,
+    searcher_generation_ids: HashSet<u64>,
+    searcher_generation_inventory: Inventory<SearcherIndexGeneration>,
+}
+
+impl WarmingStateInner {
+    /// Start tracking provided searcher as an exemplar of a new generation.
+    /// If there are active warmers, warm them with the provided searcher, and kick background GC thread if it has not yet been kicked.
+    /// Otherwise, prune state for dropped searcher generations inline.
+    fn warm_new_searcher_generation(
+        &mut self,
+        searcher: &Searcher,
+        this: &Arc<Mutex<Self>>,
+    ) -> crate::Result<()> {
+        self.searcher_generation_ids
+            .insert(searcher.index_generation().generation_id());
+        let warmers = self.pruned_warmers();
+        // Avoid threads (warming as well as background GC) if there are no warmers
+        if warmers.is_empty() {
+            return Ok(());
+        }
+        self.start_gc_thread_maybe(this)?;
+        warming_executor(self.num_warming_threads.min(warmers.len()))?
+            .map(|warmer| warmer.warm(searcher), warmers.into_iter())?;
+        Ok(())
+    }
+
+    /// Attempt to upgrade the weak Warmer references, pruning those which cannot be upgraded.
+    /// Return the strong references.
+    fn pruned_warmers(&mut self) -> Vec<Arc<dyn Warmer>> {
+        let strong_warmers = self
+            .warmers
+            .iter()
+            .flat_map(|weak_warmer| weak_warmer.upgrade())
+            .collect::<Vec<_>>();
+        self.warmers = strong_warmers.iter().map(Arc::downgrade).collect();
+        strong_warmers
+    }
+
+    /// [Warmer::garbage_collect] active warmers if some searcher generation is observed to have been dropped.
+    fn gc_maybe(&mut self) -> bool {
+        // TODO avoid GC if no gen was deleted
```

Comment:
I agree with not doing this right now, but to be more explicit "avoid GC if the only generations dropped are semantically equivalent to live generations"

Reply:
I need to drop this TODO. It is obsolete.